docs/adr/0002-project-locks.md

```diff
@@ -0,0 +1,169 @@
+	- Project represents a Terraform project. Since there may be multiple Terraform projects in a single repo we also include Path to the project root relative to the repo root.
+- [ProjectLock](https://github.com/runatlantis/atlantis/blob/main/server/events/models/models.go#L225-L240)
+	- ProjectLock represents a lock on a project.
+
+Each Repo can have many Pull Requests (One to Many)
+Each Pull Request can have many Projects (Many to Many)
+Each ProjectLock has one Project (One to One)
+
+#### Working Dir Loccks
+[Working Dir locks](https://github.com/runatlantis/atlantis/blob/main/server/events/working_dir_locker.go#L29-L52) are not part of the backend DB and thus do not have a model, instead its in-memory in `working_dir_locker`
+
+Currently can lock the entire PR or per workspace + path
+- [TryLockPull](https://github.com/runatlantis/atlantis/blob/f4fa3138d7a9dfdb494105597dce88366effff9e/server/events/working_dir_locker.go#L59-L75): Repo + PR + workspace
+- [TryLock](https://github.com/runatlantis/atlantis/blob/f4fa3138d7a9dfdb494105597dce88366effff9e/server/events/working_dir_locker.go#L77-L94): Repo + PR + workspace + path
+
+#### Stack Walk Overview
+
+Here is a high-level view of what happens when a Pull Request is opened with an auto-plan:
+
+1. Events Controller accepts POST webhook and determines which VCS client to handle the request
+2. VCSEventsController determines what type of VCS event (opened) and calls RunPlanCommand in CommandRunner
+3. CommandRunner validates the context and runs pre-hooks (if they exist)
+    1. WorkingDirLocker locks - `w.WorkingDirLocker.TryLock(baseRepo.FullName, pull.Num, DefaultWorkspace, DefaultRepoRelDir)`
+    2. Git Repo is cloned - `w.WorkingDir.Clone(log, headRepo, pull, DefaultWorkspace)`
+    3. run hooks
+    4. WorkingDir lock is released when the pre-hook function returns
+4. CommandRunner determines which command runner to use (Plan)
+5. PlanCommandRunner determines projects affected by the Pull Request by calling projectCmdBuilder 
+    1. VCS client returns modified files
+    2. returns here/skips cloning in some cases `--skip-clone-no-changes`
+    3. WorkingDirLocker locks - `p.WorkingDirLocker.TryLock(ctx.Pull.BaseRepo.FullName, ctx.Pull.Num, workspace, DefaultRepoRelDir)`
+    4. Git Repo is cloned - `p.WorkingDir.Clone(ctx.Log, ctx.HeadRepo, ctx.Pull, workspace)`
+    5. parses server and repo configs
+    6. determines projects and returns them
+    7. WorkingDir lock is released
+6. PlanCommandRunner cleans up previous plans and Project locks
+7. PlanCommandRunner passes ProjectCommandRunner and a list of projects to ProjectCommandPoolExecutor which executes  `ProjectCommandRunner.doPlan`
+8. ProjectCommandRunner.doPlan 
+	1. acquires Project lock - `p.Locker.TryLock(ctx.Log, ctx.Pull, ctx.User, ctx.Workspace, models.NewProject(ctx.Pull.BaseRepo.FullName, ctx.RepoRelDir), ctx.RepoLocking)`
```

Comment:
Moving this project lock up into projectCmdBuilder before step 5.3 (workingdirlocker locks) would mean we could give instant feedback about the lock problem instead of possibly hours later when we get around to the plan step that fails to get the lock, and we could drop the clone here in doPlan step 3.

Reply:
Do we really want to lock on the whole PR so early? There's on going work in #3879 to an the ability to move it even deeper to the apply step instead of at the first plan.